source/structured-streaming.txt

```diff
@@ -0,0 +1,327 @@
+
+   To learn more about change streams, see 
+   :manual:`Change Streams </changeStreams>` in the MongoDB manual.
+
+Specify read stream configuration settings on your local 
+SparkSession ``.readStream``. Lastly, add ``.load()`` to create a read 
+stream with your configuration.
+
+.. note:: 
+   
+   Spark does not begin streaming until you call the ``.start()`` 
+   method on a streaming query.
+
+You must specify the following configuration settings to read from 
+MongoDB:
+
+.. list-table::
+   :header-rows: 1
+   :stub-columns: 1
+   :widths: 10 40
+
+   * - Setting
+     - Description
+
+   * - ``.readStream.format()``
+     - The format to use for read stream data. Use "mongodb".
+
+   * - ``.writeStream.trigger()``
+     - Enables continuous processing for your ``.readStream``. Use 
+       the ``continuous`` parameter.
+
+.. tabs-drivers::
+
+   tabs:
+     - id: java-sync
+       content: |
+
+     - id: python
+       content: |
+
+         .. code-block:: python
+            :copyable: true
+            :emphasize-lines: 3, 9
+         
+            streamingDataFrame = (<local SparkSession>
+              .readStream
+              .format("mongodb")
+              .load()
+            )
+         
+            query = (streamingDataFrame
+              .writeStream
+              .trigger(continuous="1 second")
+              .format("memory")
+              .outputMode("append")
+            )
+
+            query.start()
+
+         For a complete list of methods, see the 
+         `pyspark Structured Streaming reference <https://spark.apache.org/docs/latest/api/python/reference/pyspark.ss.html>`__.
+
+     - id: scala
+       content: |
+
+Examples
+--------
+
+The following examples show Spark Structured Streaming configurations 
+for streaming between MongoDB and a CSV file.
+
+Stream to MongoDB from a CSV File
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. tabs-drivers::
+
+   tabs:
+     - id: java-sync
+       content: |
+
+         .. code-block:: java
+            :copyable: true
+
+     - id: python
+       content: |
+
+         To create a :ref:`write stream <write-structured-stream>` to 
+         MongoDB from a ``.csv`` file, first create a `DataStreamReader <https://spark.apache.org/docs/latest/api/python/reference/api/pyspark.sql.streaming.DataStreamReader.html>`__ 
+         from the ``.csv`` file, then use that ``DataStreamReader`` to 
+         create a `DataStreamWriter <https://spark.apache.org/docs/latest/api/python/reference/api/pyspark.sql.streaming.DataStreamWriter.html>`__ 
+         to MongoDB. Finally, use the ``.start()`` method to begin the 
+         stream.
+         
+         As streaming data is read from the ``.csv`` file, it is added 
+         to MongoDB in the `outputMode <https://spark.apache.org/docs/latest/api/python/reference/api/pyspark.sql.streaming.DataStreamWriter.outputMode.html#pyspark.sql.streaming.DataStreamWriter.outputMode>`__ 
+         you specify.
```

Comment:
Suggestion: I'm not sure the provided description adds value to the code snippet outside of presenting the links. I think the paragraph largely duplicates the content of the code. Consider moving the links to after the code snippet and simplifying the introduction.

If accepted, this also applies to the following example.

```suggestion
        The following code snippet demonstrates how to stream data from a `.csv` file to MongoDB:
        
        <code snippet>
        
        For more information on the classes used in the preceding code, see the following resources:
        
        - <list of API documentation links>
```

Reply:
I'm not opposed to hand-holding (I think the more help we can give the better), what I think could be improved about the existing text is that the density of class names and links makes it hard to understand. In addition, `DataStream[Reader/Writer]` were not discussed earlier on the page, so this paragraph introduces new terms which a reader may not need to understand to understand what the code is doing.

I think the text obfuscates the task the code is accomplishing ("write data from MongoDB to a CSV" becomes "create a DataStreamReader to create a DataStreamWriter to MongoDB") rather than clarifies it.

I think focusing on "what the code is doing" rather than "what programming constructs the code is using" will help readers better find the answers to their questions.

If you disagree with this assessment, an alternative way to break this up could be to organize this paragraph as an ordered list which discusses the tasks done be the code sequentially (however I feel this still contains the issue of discussing classes which a reader does not need to be made aware of to achieve their goal).